eclair-core/src/main/scala/fr/acinq/eclair/db/pg/PgUtils.scala

```diff
@@ -18,81 +18,174 @@ package fr.acinq.eclair.db.pg
+      override def obtainExclusiveLock(implicit ds: DataSource): Unit =
+        obtainDatabaseLease(instanceId, leaseDuration)
+
+      override def withLock[T](f: Connection => T)(implicit ds: DataSource): T = {
+        inTransaction { connection =>
+          val res = f(connection)
+          checkDatabaseLease(connection, instanceId, lockExceptionHandler)
+          res
+        }
+      }
 
-  sealed trait DatabaseLock {
-    def obtainExclusiveLock(implicit ds: DataSource): Unit
+      private def obtainDatabaseLease(instanceId: UUID, leaseDuration: FiniteDuration, attempt: Int = 1)(implicit ds: DataSource): Unit = synchronized {
+        logger.debug(s"trying to acquire database lease (attempt #$attempt) instance ID=$instanceId")
+
+        if (attempt > 3) throw new TooManyLockAttempts("Too many attempts to acquire database lease")
+
+        try {
+          inTransaction { implicit connection =>
+            acquireExclusiveTableLock()
+            getCurrentLease match {
+              case Some(lease) =>
+                if (lease.instanceId == instanceId || lease.expired)
+                  updateLease(instanceId, leaseDuration)
+                else
+                  throw new LeaseException(s"The database is locked by instance ID=${lease.instanceId}")
+              case None =>
+                updateLease(instanceId, leaseDuration, insertNew = true)
+            }
+          }
+          logger.debug("database lease was successfully acquired")
+        } catch {
+          case e: PSQLException if e.getServerErrorMessage != null && e.getServerErrorMessage.getSQLState == "42P01" =>
+            withConnection {
+              connection =>
+                logger.warn(s"table $LeaseTable does not exist, trying to recreate it")
+                initializeLeaseTable(connection)
+                obtainDatabaseLease(instanceId, leaseDuration, attempt + 1)
+            }
+        }
+      }
 
-    def withLock[T](f: Connection => T)(implicit ds: DataSource): T
-  }
+      private def initializeLeaseTable(implicit connection: Connection): Unit = {
+        using(connection.createStatement()) {
+          statement =>
+            // allow only one row in the ownership lease table
+            statement.executeUpdate(s"CREATE TABLE IF NOT EXISTS $LeaseTable (id INTEGER PRIMARY KEY default(1), expires_at TIMESTAMP NOT NULL, instance VARCHAR NOT NULL, CONSTRAINT one_row CHECK (id = 1))")
+        }
+      }
 
-  case object NoLock extends DatabaseLock {
-    override def obtainExclusiveLock(implicit ds: DataSource): Unit = ()
+      private def acquireExclusiveTableLock()(implicit connection: Connection): Unit = {
+        using(connection.createStatement()) {
+          statement =>
+            statement.executeUpdate(s"SET lock_timeout TO '${LockTimeout.toSeconds}s'")
```

Comment:
@rorp. Why not use `NOWAIT`  here?

Reply:
I went ahead and changed it in 723d926, but I'm still interested in your rationale opting for a non-zero timeout.

Reply:
AFAIR there was some weirdness in exceptions Pg driver threw in NOWAIT mode. Should work in both ways.

Reply:
Seeing that lock errors seem to be handled fine in tests, I think we can go ahead with `NOWAIT`.

Reply:
Okay I think I know why it was done that way. There can be a conflict between a regular SQL requests checking the lock, and the background task that regularly renew the lease. If those happen at the same time, one of them will fail which will make the app stop. Adding a 5 seconds timeout provided some leeway, one of the lock attempts would just wait after the first one is completed.